### PR TITLE
Added stopping check for callbacks

### PR DESCRIPTION
**Pull request by haoxiangyang89**

### CODE REVIEW

**Comment by joehuchette** [comment]
Looks great, thanks! Could you add tests for the new behavior, similar to this: https://github.com/JuliaOpt/JuMP.jl/blob/master/test/old/callback.jl#L295-L305

**Comment by mlubin** [comment]
These tests may be flakey for the callbacks that aren't required to be called (cut callback, heur callback). We're just testing the solver and not really JuMP in that case. I'd be okay to merge just with a test on the info callback.

**Comment by joehuchette** [comment]
Looks like CI failures:
```
Callbacks: Error During Test
  Got an exception of type UndefVarError outside of a @test
  UndefVarError: solver not defined
  Stacktrace:
   [1] macro expansion at .\test.jl:917 [inlined]
   [2] macro expansion at C:\Users\appveyor\.julia\v0.6\JuMP\test\callback.jl:307 [inlined]
   [3] macro expansion at .\test.jl:860 [inlined]
   [4] anonymous at .\<missing>:?
   [5] include_from_node1(::String) at .\loading.jl:576
   [6] include at .\sysimg.jl:14
   [7] include_from_node1 at .\loading.jl:576
   [8] include at .\sysimg.jl:14
   [9] process_options at .\client.jl:305
   [10] _start at .\client.jl:371
```

File: test/callback.jl [hunk 1]
```diff
@@ -304,6 +304,40 @@ using Base.Test
         @test solve(mod, suppress_warnings=true) == :UserLimit
     end
 
+    @testset "Callback exit on return StopTheSolver with $solver" for infosolver in info_solvers
```
Comment (joehuchette) [inline test/callback.jl]: should be ``"Callback exit on return StopTheSolver with $infosolver"``

File: test/callback.jl [hunk 2]
```diff
@@ -304,6 +304,40 @@ using Base.Test
...
+        mycallback = _ -> JuMP.StopTheSolver
+        addinfocallback(mod, mycallback, when = :MIPSol)
+        @test solve(mod, suppress_warnings=true) == :UserLimit
+
+        nodes      = Int[]
+        entered = fill(false, 2)
+
+        N = 10000
+        include(joinpath("data","informational.jl"))
+        mod = Model(solver=GurobiSolver())
```
Comment (joehuchette) [inline test/callback.jl]: Should be ``infosolver`` instead of ``GurobiSolver()``

File: test/callback.jl (hunk 1 above)
Comment (haoxiangyang89) [reply test/callback.jl]: Both are corrected now. Sorry I was not very experienced with @test structure.

**Comment by haoxiangyang89** [comment]
It seems that it will not pass the CPLEX test because of the test instance. CPLEX does not obtain a Intermediate state. I will fix this and upload a test again.

File: test/callback.jl (hunk 1 above)
Comment (joehuchette) [reply test/callback.jl]: No apologies necessary, the JuMP test currently a bit difficult to work with.

File: test/callback.jl [hunk 3]
```diff
@@ -304,6 +304,40 @@ using Base.Test
...
+            @assert MathProgBase.cbgetstate(cb) == :Intermediate
+            entered[1] = true
+            push!(nodes,      MathProgBase.cbgetexplorednodes(cb))
+            if MathProgBase.cbgetexplorednodes(cb) >= 2
+                return JuMP.StopTheSolver
+            end
+        end
+        addinfocallback(mod, myinfo1, when = :Intermediate)
+        @test solve(mod, suppress_warnings=true) == :UserLimit
+        @test maximum(nodes) >= 2
```
Comment (mlubin) [inline test/callback.jl]: Shouldn't the test be for `<= 2`?

File: test/callback.jl [hunk 4]
```diff
@@ -304,6 +304,40 @@ using Base.Test
...
+        include(joinpath("data","informational.jl"))
+        mod = Model(solver=infosolver)
+        @variable(mod, x[1:N], Bin)
+        @objective(mod, Max, dot(r1,x))
+        @constraint(mod, c[i=1:10], dot(r2[i],x) <= rhs[i]*N/10)
+        # Test that solver fills solution correctly
+        function myinfo1(cb)
+            @assert MathProgBase.cbgetstate(cb) == :Intermediate
+            entered[1] = true
+            push!(nodes,      MathProgBase.cbgetexplorednodes(cb))
```
Comment (mlubin) [inline test/callback.jl]: Weird extra spacing here

File: test/callback.jl [hunk 5]
```diff
@@ -304,6 +304,40 @@ using Base.Test
...
+        @variable(mod, 0 <= x <= 2, Int)
+        @variable(mod, 0 <= y <= 2, Int)
+        @objective(mod, Max, x + 2y)
+        @constraint(mod, y + x <= 3.5)
+
+        mycallback = _ -> JuMP.StopTheSolver
+        addinfocallback(mod, mycallback, when = :MIPSol)
+        @test solve(mod, suppress_warnings=true) == :UserLimit
+
+        nodes      = Int[]
```
Comment (mlubin) [inline test/callback.jl]: weird extra spacing here

**Comment by mlubin** [comment]
I merged the fix without the tests into `release-0.18`. Doesn't matter too much since it's a dead-end branch.

**Pull request closed without merging**